src/toxcore/dht/server/mod.rs

```diff
@@ -93,136 +105,250 @@ pub struct Server {
-    /// Close List (contains nodes close to own DHT PK)
-    pub kbucket: Kbucket,
-}
-
 impl Server {
     /**
     Create new `Server` instance.
     */
     pub fn new(tx: Tx, pk: PublicKey, sk: SecretKey) -> Server {
-        let kbucket = Kbucket::new(&pk);
-
         debug!("Created new Server instance");
         Server {
             sk,
             pk,
             tx,
-            state: Arc::new(RwLock::new(ServerState {
-                peers_cache: HashMap::new(),
-                kbucket,
-            })),
+            is_hole_punching_enabled: true,
+            peers_cache: Arc::new(RwLock::new(HashMap::new())),
+            close_nodes: Arc::new(RwLock::new(Kbucket::new(&pk))),
             onion_symmetric_key: Arc::new(RwLock::new(new_symmetric_key())),
             onion_announce: Arc::new(RwLock::new(OnionAnnounce::new(pk))),
+            friends: Arc::new(RwLock::new(Vec::new())),
+            bootstrap_nodes: Arc::new(RwLock::new(Bucket::new(None))),
+            bootstrap_times: Arc::new(RwLock::new(0)),
+            last_nodes_req_time: Arc::new(RwLock::new(Instant::now())),
             tcp_onion_sink: None
         }
     }
+
+    /// return peers_cache member variable
+    pub fn get_peers_cache(&self) -> &Arc<RwLock<HashMap<PublicKey, ClientData>>> {
+        &self.peers_cache
+    }
+
+    /// add friend
+    pub fn add_friend(&self, friend: DhtFriend) -> IoFuture<()> {
+        let mut friends = self.friends.write();
+
+        friends.push(friend);
+
+        Box::new(future::ok(()))
+    }
+
+    /// main loop of dht server, call this function every second
+    pub fn dht_main_loop(&self, kill_timeout: u64, ping_timeout: u64, ping_interval: u64, bad_node_timeout: u64, nodes_req_interval: u64) -> IoFuture<()> {
+        self.remove_timedout_clients(Duration::from_secs(kill_timeout));
+        self.remove_timedout_ping_ids(Duration::from_secs(ping_timeout));
+
+        let mut friends = self.friends.write();
+
+        let ping_bootstrap_nodes = self.ping_bootstrap_nodes();
+        let ping_and_get_close_nodes = self.ping_and_get_close_nodes(ping_interval);
+        let send_nodes_req_random = self.send_nodes_req_random(bad_node_timeout, nodes_req_interval);
+        let send_nodes_req_packets = {
+            let nodes_sender = friends.iter_mut()
+                .map(|friend| {
+                    friend.send_nodes_req_packets(self, PING_INTERVAL, NODES_REQ_INTERVAL, BAD_NODE_TIMEOUT)
+                });
+
+            let nodes_stream = stream::futures_unordered(nodes_sender).then(|_| Ok(()));
+
+            nodes_stream.for_each(|()| Ok(()))
+        };
+
+        let res = ping_bootstrap_nodes.join4(ping_and_get_close_nodes,
+                                             send_nodes_req_random,
+                                             send_nodes_req_packets)
+            .map(|((), (), (), ())| ());
+
+        self.bootstrap_nodes.write().deref_mut().nodes.clear();
```

Comment:
If you move this to `ping_bootstrap_nodes` function you won't need to obtain lock twice.

Reply:
done